### PR TITLE
Make v1model immutable

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
In particular, v1model::instance::version was a global static variable. That's bad.
The weird thing is that some of the classes I have introduced to represent the two versions of v1model are currently not used.
@antoninbas some of these are in the control-plane API generation code.
In particular, V1MODEL2020 is never used, but all tests pass.

### CODE REVIEW

File: midend/interpreter.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ bool SymbolicValueFactory::isFixedWidth(const IR::Type* type) const {
 }
 
 unsigned SymbolicValueFactory::getWidth(const IR::Type* type) const {
-    type = typeMap->getType(type, true);
+    type = typeMap->getTypeType(type, true);
```
Comment (fruffy) [inline midend/interpreter.cpp:82]: This seems unrelated to the rest of the changes. What is this for?

File: midend/interpreter.cpp [hunk 2]
```diff
@@ -739,7 +739,7 @@ void ExpressionEvaluator::postorder(const IR::Operation_Relation* expression) {
 
 void ExpressionEvaluator::postorder(const IR::Member* expression) {
     auto type = typeMap->getType(expression, true);
-    if (type->is<IR::Type_MethodBase>()) {
+    if (type->is<IR::Type_MethodBase>() || type->is<IR::Type_Error>()) {
```
Comment (fruffy) [inline midend/interpreter.cpp:742]: Same here.

File: test/gtest/helpers.cpp [hunk 3]
```diff
@@ -128,7 +128,7 @@ P4CTestEnvironment::P4CTestEnvironment() {
     // XXX(seth): We should find a more robust way to locate these headers.
     _coreP4 = readHeader("p4include/core.p4");
     _v1Model = readHeader("p4include/v1model.p4", true,
-                          "V1MODEL_VERSION", P4V1::V1Model::instance.version);
+                          "V1MODEL_VERSION", 20200408);
```
Comment (fruffy) [inline test/gtest/helpers.cpp:131]: Shouldn't this be using getV1ModelVersion?

**Review by fruffy — APPROVED** [review]
I do not know much about how the v1model versioning works or what it is used for. So my feedback on this is only high level.

File: frontends/p4/uniqueNames.h [hunk 4]
```diff
@@ -80,8 +80,14 @@ class FindSymbols : public Inspector {
     }
     void postorder(const IR::Declaration_Variable* decl) override
     { doDecl(decl); }
-    void postorder(const IR::Declaration_Constant* decl) override
-    { doDecl(decl); }
+    void postorder(const IR::Declaration_Constant* decl) override {
+        // Skip toplevel constants with names like __
+        // We assume that these do not clash and no new symbols with
+        // these names will be added.
+        if (decl->getName().name.startsWith("__") && getParent<IR::P4Program>())
```
Comment (fruffy) [inline frontends/p4/uniqueNames.h:87]: Could this cause an issue with inlining and constants that have been declared with __?

**Comment by hesingh** [comment]
The original intent to support more than one version of v1model.p4 makes sense to me.  How we support is another question.

File: midend/interpreter.cpp (hunk 1 above)
Comment (mbudiu-vmw) [reply midend/interpreter.cpp:82]: Yes, this is correct. This file is in fact not currently used, but these changes were triggered by @VolodymyrPeschanenkoLitSoft
and his changes to the loop unrolling pass. This file is part of loop unrolling, and these changes are indeed fixing some unrelated bugs in this file.

File: test/gtest/helpers.cpp (hunk 3 above)
Comment (mbudiu-vmw) [reply test/gtest/helpers.cpp:131]: No, getV1ModelVersion extracts the version from an existing loaded program, but here we haven't loaded any program yet. What we are doing is preparing to load the program using this specific version of v1model.

File: frontends/p4/uniqueNames.h (hunk 4 above)
Comment (mbudiu-vmw) [reply frontends/p4/uniqueNames.h:87]: I don't think we ever move objects into the global namespace currently, so I don't think this is an issue with the current implementation. You are right that in the future this may cause problems. The right solution is to reserve some names in the spec for internal use of the compiler, and then we can guarantee that this is safe. I could make this safer by checking explicitly just for __v1model_version, although that won't work for other architectures. Maybe we should discuss this in the LDWG.

**Comment by fruffy** [comment]
Sounds good. From my point, I have no further comments.

File: test/gtest/helpers.cpp (hunk 3 above)
Comment (fruffy) [reply test/gtest/helpers.cpp:131]: I do not quite understand the versioning. Where does 20200408 come from? Last change to the v1model.p4 file?

**Comment by mbudiu-vmw** [comment]
> 
> 
> I do not know much about how the v1model versioning works or what it is used for. So my feedback on this is only high level.

If you look at the file v1mode.p4 it starts this way: 

```
#ifndef V1MODEL_VERSION
#define V1MODEL_VERSION 20180101
#endif

const bit<32> __v1model_version = V1MODEL_VERSION;

#if V1MODEL_VERSION >= 20200408
typedef bit<9>  PortId_t;       // should not be a constant size?
#endif
```

So we have in fact two versions of v1model.p4 in the same file, depending on the macro V1MODEL_VERSION when you include the file. Currently there are only 2 versions, but the file can evolve further in the future.

When we compile P4-14 programs we convert them to P4-16 programs that need to use the 20200408 version, so that's why you see the macro being defined this way before invoking the preprocessor. This behavior is identical prior to this refactoring, the V1Model class had a default value for the version field of 20200408.

The problem we are fixing in this PR is that the field V1Model::instance.version was a mutable global, which could be used to carry information outside the IR or pass arguments, which is not a good idea if you want the code to be readable and especially thread-safe. We still have a V1Model::instance global, but now it's a constant. To figure out which version of v1model was used when compiling a program you have to look at the __v1model_instance global constant, which a visitor pass provides.

**Comment by mbudiu-vmw** [comment]
I will merge this PR to unblock #2687

File: frontends/p4/fromv1.0/programStructure.cpp [hunk 5]
```diff
@@ -609,7 +609,7 @@ void ProgramStructure::include(cstring filename, cstring ppoptions) {
 void ProgramStructure::loadModel() {
     // This includes in turn core.p4
     std::stringstream versionArg;
-    versionArg << "-DV1MODEL_VERSION=" << V1Model::instance.version;
+    versionArg << "-DV1MODEL_VERSION=20200408";
```
Comment (ChrisDodd) [inline frontends/p4/fromv1.0/programStructure.cpp:612]: It would be better to leave V1Model::instance.version as a static constexpr in the class, rather than hardcoding this value here.

**Comment by ChrisDodd** [comment]
> In particular, v1model::instance::version was a global static variable. That's bad.

I disagree with this -- having a global static constant is fine.  It might be questionable if it was mutable, but it is not.

**Comment by mbudiu-vmw** [comment]
I was mutated during compilation.

**Comment by mbudiu-vmw** [comment]
We can have two constexpr values for the two values of v1model that we support. But we should not have a global mutable static.
